torch/csrc/inductor/static_cuda_launcher.cpp

```diff
@@ -0,0 +1,462 @@
+#if defined(USE_CUDA) && !defined(USE_ROCM)
+// We disable this file from being hipified because there are CUDA drivers hip
+// has not implemented yet. Also, we're passing in a cubin file directly, so it
+// would take more work to support ROCM anyway.
+#include <torch/csrc/utils/pythoncapi_compat.h>
+
+#include <ATen/Context.h>
+#include <ATen/cuda/Exceptions.h>
+#include <ATen/cuda/nvrtc_stub/ATenNVRTC.h>
+#include <c10/cuda/CUDAGuard.h>
+#include <c10/cuda/CUDAStream.h>
+#include <torch/csrc/inductor/static_cuda_launcher.h>
+#include <cstdint>
+#include <stdexcept>
+
+#include <torch/csrc/utils/python_numbers.h>
+#include <filesystem>
+#include <optional>
+/**
+  Implements a static launcher for triton compiled CUDA kernels.
+  Given a path to a cubin file, a function name, and some metadata,
+  this class loads and launches the cubin.
+
+  Doing this avoids C++ codegen and compilation during compile, since we can
+  use a statically compiled library to launch the kernel. To avoid mallocing
+  for the arguments, we have a launcher for different numbers of arguments up
+  to a max. StaticCudaLauncher only supports # of arguments up until 10 for
+  now.
+
+  Note that we allocate 8 bytes per argument, no matter the types of each
+  argument, since we don't know ahead of time what the types of each argument
+  passed to the triton kernel are. This may take slightly more memory on the
+  stack, and will require some benchmarking. However, since the vast majority
+  of triton kernels have less than 10 args, this seems unlikely to be
+  expensive.
+
+  This launcher is paired with StaticallyLaunchedCudaKernel in
+  triton_heuristics.py.
+
+  TODO:
+  - Handle CutensorMap, NvtmDesc
+  - Handle launch_enter and launch_exit hooks (in python maybe?)
+ */
+
+// Use ATen/NVRTC.h to gain access to the CUDA driver API.
+// This function is only called when CUDA is enabled, and only called to load
+// and launch triton compiled CUDA kernels, so CUDA should always be
+// initialized.
+static const at::cuda::NVRTC& nvrtc() {
+  return at::globalContext().getNVRTC();
+}
+
+static CUdeviceptr getPointer(PyObject* obj) {
+  CUdeviceptr data_ptr = 0;
+  if (THPUtils_checkLong(obj)) {
+    data_ptr = THPUtils_unpackUInt64(obj);
+    return data_ptr;
+  }
+  if (obj == Py_None) {
+    // valid nullptr
+    return data_ptr;
+  }
+  auto ptr = THPObjectPtr{PyObject_GetAttrString(obj, "data_ptr")};
+  if (ptr) {
+    auto empty_tuple = THPObjectPtr{PyTuple_New(0)};
+    auto ret = THPObjectPtr{PyObject_Call(ptr, empty_tuple, nullptr)};
+    if (!THPUtils_checkLong(ret)) {
+      throw std::runtime_error(
+          "data_ptr method of Pointer object must return 64-bit int");
+    }
+    data_ptr = THPUtils_unpackUInt64(ret);
```

Comment:
Empty tensors will have data-ptr of 0:
```
>>> import torch
>>> torch.tensor([]).data_ptr()
0
```

I could imagine this happening with a dynamic shape kernel. Can we handle this / add test ?

Reply:
i guess the current triton logic does something very similar.. https://github.com/triton-lang/triton/blob/18b8d1d0f44cd641b0ac6f041bda5a7d747b9acc/third_party/nvidia/backend/driver.py#L313-L357

Reply:
Hmm can you think of a way to write a triton kernel that does something meaningful (i.e. not constant) to an empty tensor? You can't really load from it because that block of memory is empty, right? I'm having trouble coming up with one that compiles. 

Or maybe there's a torch.compile() example and I could tlparse it to see what triton kernel it generates?

Reply:
Here is an example program:

```
import torch

@torch.compile()
def foo(x, y):
    return torch.cat(((x * 4), y + 10))

y = torch.rand(20, device="cuda")
t = torch.rand(10, device="cuda")
torch._dynamo.decorators.mark_unbacked(t, 0)
foo(t, y)

t = torch.rand(0, device="cuda")
torch._dynamo.decorators.mark_unbacked(t, 0)
foo(t, y)
```

Reply:
You're not loading from the empty tensor in the second run but the kernel is doing something meaningful..